src/imitation/algorithms/bc.py

```diff
@@ -114,9 +120,23 @@ def __call__(
             A BCTrainingMetrics object with the loss and all the components it
             consists of.
         """
-        obs = util.safe_to_tensor(obs)
+        tensor_obs: Union[th.Tensor, Dict[str, th.Tensor]]
+        if isinstance(obs, types.DictObs):
+            tensor_obs = {k: util.safe_to_tensor(v) for k, v in obs.unwrap().items()}
+        elif isinstance(obs, dict):
+            tensor_obs = {k: util.safe_to_tensor(v) for k, v in obs.items()}
+        else:
+            tensor_obs = util.safe_to_tensor(obs)
         acts = util.safe_to_tensor(acts)
-        _, log_prob, entropy = policy.evaluate_actions(obs, acts)
+
+        # policy.evaluate_actions's type signature seems wrong to me.
```

Comment:
I think `evaluate_action` is assuming that `extract_features` is `th.Tensor -> th.Tensor` like the default implementation (although I agree you could override it).

`obs_to_tensor` seems to be resopnsible for dict flattening: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/policies.py#L231 (this may contradict information I told you in the past, I wrongly thought feature extractors did this, I think they maybe used to in older versions of SB3?)

Reply:
`obs_to_tensor` claims to flatten dicts, but it doesn't

```
from gym import spaces
from stable_baselines3.common import policies

obs_space = spaces.Dict({"a": spaces.Discrete(2,), "b": spaces.Box(0,1,(2,))})
policy = policies.MultiInputActorCriticPolicy(
    obs_space,
    spaces.Discrete(4,),
    lambda _: 0.001,
)
obs = obs_space.sample()
policy.obs_to_tensor(obs)
```
gives `({'a': tensor([1]), 'b': tensor([[0.7694, 0.0154]])}, False)`

Reply:
Probably worth making an issue in SB3, I didn't see any previous discussion when I looked earlier

Reply:
Oh, huh, well spotted -- yes, we should open an issue in SB3 and have them fix their type annotations, this is very confusing right now.

Reply:
Suggest making that issue and converting this to a TODO linking to that issue.

Reply:
https://github.com/DLR-RM/stable-baselines3/issues/1679